README.md
# `pe_info`

Small naive utility to detect which MSVC version/Windows version was used to compile an executable.

This utility relies almost entirely on self reported data it is not accurate on executables where the author might want to obfuscate the origin.

# Build

1. Install [rust](https://www.rust-lang.org/tools/install).
2. Clone the repository with `git clone https://github.com/gtker/pe_info`.
3. Run with `cargo run --release -- [OPTIONS] [FILES]`, or build with `cargo build --release` and move the binary from `./target` to your `PATH`.

# Usage

Grab the [newest release](https://github.com/gtker/pe_info/releases).

```
Usage: pe_info [OPTIONS] [FILES]...

Arguments:
  [FILES]...  Files to analyze

Options:
  -q, --quiet    Do not print out intermediate results
  -s, --strict   Return error code for likely invalid information
  -h, --help     Print help
  -V, --version  Print version
```

Usage on some Warcraft 3 executables:

```text
$ pe_info War3_1.13.exe Storm_1.13.dll Game_1.13.dll game.dll Storm.dll War3_1.0.exe
Reading 'War3_1.13.exe'
[ Self Reported ] Timestamp is 2003-05-18 23:44:04 UTC, which means the newest possible version is Visual Studio .NET 2003 (24 days after release)
[ Heuristic     ] Links against msvcrt.dll which suggests compiler is either Visual C++ 4.2 (valid for timestamp), Visual C++ 5.0 (Visual Studio 97) (valid for timestamp), Visual C++ 6.0 (Visual Studio 6.0) (valid for timestamp)
[ Heuristic     ] Imported symbol '??1type_info@@UAE@XZ' from 'msvcrt.dll' suggests C++ was used.
[ Self Reported ] Linker version is 6.0, which suggests linker is Visual C++ 6.0 (Visual Studio 6.0)
[ Self Reported ] Operating system version is 4.0 which suggests Windows 95
[ Self Reported ] Subsystem version 4.0 agrees with operating system version of Windows 95
Compiled: 2003-05-18 23:44:04 UTC
Compiler (one of):
         Visual C++ 4.2
         Visual C++ 5.0 (Visual Studio 97)
         Visual C++ 6.0 (Visual Studio 6.0)
Linker: Visual C++ 6.0 (Visual Studio 6.0)
Operating System: Windows 95

Reading 'Storm_1.13.dll'
[ Self Reported ] Timestamp is 2003-05-18 23:43:38 UTC, which means the newest possible version is Visual Studio .NET 2003 (24 days after release)
[ Heuristic     ] Links against MSVCRT.dll which suggests compiler is either Visual C++ 4.2 (valid for timestamp), Visual C++ 5.0 (Visual Studio 97) (valid for timestamp), Visual C++ 6.0 (Visual Studio 6.0) (valid for timestamp)
[ Heuristic     ] Imported symbol '??1type_info@@UAE@XZ' from 'MSVCRT.dll' suggests C++ was used.
[ Self Reported ] Linker version is 6.0, which suggests linker is Visual C++ 6.0 (Visual Studio 6.0)
[ Self Reported ] Operating system version is 4.0 which suggests Windows 95
[ Self Reported ] Subsystem version 4.0 agrees with operating system version of Windows 95
Compiled: 2003-05-18 23:43:38 UTC
Compiler (one of):
         Visual C++ 4.2
         Visual C++ 5.0 (Visual Studio 97)
         Visual C++ 6.0 (Visual Studio 6.0)
Linker: Visual C++ 6.0 (Visual Studio 6.0)
Operating System: Windows 95

Reading 'Game_1.13.dll'
[ Self Reported ] Timestamp is 2003-05-18 23:57:48 UTC, which means the newest possible version is Visual Studio .NET 2003 (24 days after release)
[ Heuristic     ] Links against MSVCRT.dll which suggests compiler is either Visual C++ 4.2 (valid for timestamp), Visual C++ 5.0 (Visual Studio 97) (valid for timestamp), Visual C++ 6.0 (Visual Studio 6.0) (valid for timestamp)
[ Heuristic     ] Imported symbol '??1type_info@@UAE@XZ' from 'MSVCRT.dll' suggests C++ was used.
[ Self Reported ] Linker version is 6.0, which suggests linker is Visual C++ 6.0 (Visual Studio 6.0)
[ Self Reported ] Operating system version is 4.0 which suggests Windows 95
[ Self Reported ] Subsystem version 4.0 agrees with operating system version of Windows 95
Compiled: 2003-05-18 23:57:48 UTC
Compiler (one of):
         Visual C++ 4.2
         Visual C++ 5.0 (Visual Studio 97)
         Visual C++ 6.0 (Visual Studio 6.0)
Linker: Visual C++ 6.0 (Visual Studio 6.0)
Operating System: Windows 95

Reading 'game.dll'
[ Self Reported ] Timestamp is 2016-12-09 06:05:18 UTC, which means the newest possible version is Visual Studio 2015 (538 days after release)
[ Heuristic     ] Links against MSVCR120.dll which suggests compiler is Visual Studio 2013, which is possible with the release date of the compiler
[ Heuristic     ] Links against MSVCP120.dll which suggests compiler is Visual Studio 2013, which is possible with the release date of the compiler
[ Heuristic     ] Imported symbol '?SCreateProcess@@YIHPBDPAD@Z' from 'Storm.dll' suggests C++ was used.
[ Heuristic     ] Imported symbol '?terminate@@YAXXZ' from 'MSVCR120.dll' suggests C++ was used.
[ Heuristic     ] Imported symbol '?_Syserror_map@std@@YAPBDH@Z' from 'MSVCP120.dll' suggests C++ was used.
[ Heuristic     ] Import of symbol 'strtoull' from 'MSVCR120.dll' suggests at least standard C++11
[ Heuristic     ] Import of symbol 'strtoll' from 'MSVCR120.dll' suggests at least standard C++11
[ Self Reported ] Linker version is 12.0, which suggests linker is Visual Studio 2013
[ Self Reported ] Operating system version is 5.1 which suggests Windows XP
[ Self Reported ] Subsystem version 5.1 agrees with operating system version of Windows XP
Compiled: 2016-12-09 06:05:18 UTC
Compiler: Visual Studio 2013
Linker: Visual Studio 2013
C++ Standard used: At least C++11
Operating System: Windows XP

Reading 'Storm.dll'
[ Self Reported ] Timestamp is 2016-12-09 06:05:51 UTC, which means the newest possible version is Visual Studio 2015 (538 days after release)
[ Heuristic     ] Links against MSVCR120.dll which suggests compiler is Visual Studio 2013, which is possible with the release date of the compiler
[ Heuristic     ] Imported symbol '?terminate@@YAXXZ' from 'MSVCR120.dll' suggests C++ was used.
[ Self Reported ] Linker version is 12.0, which suggests linker is Visual Studio 2013
[ Self Reported ] Operating system version is 5.1 which suggests Windows XP
[ Self Reported ] Subsystem version 5.1 agrees with operating system version of Windows XP
Compiled: 2016-12-09 06:05:51 UTC
Compiler: Visual Studio 2013
Linker: Visual Studio 2013
Operating System: Windows XP

Reading 'War3_1.0.exe'
[ Self Reported ] Timestamp is 2002-06-06 11:25:18 UTC, which means the newest possible version is Visual Studio .NET 2002 (113 days after release)
[ Heuristic     ] Links against msvcrt.dll which suggests compiler is either Visual C++ 4.2 (valid for timestamp), Visual C++ 5.0 (Visual Studio 97) (valid for timestamp), Visual C++ 6.0 (Visual Studio 6.0) (valid for timestamp)
[ Heuristic     ] Imported symbol '??1type_info@@UAE@XZ' from 'msvcrt.dll' suggests C++ was used.
[ Self Reported ] Linker version is 6.0, which suggests linker is Visual C++ 6.0 (Visual Studio 6.0)
[ Self Reported ] Operating system version is 4.0 which suggests Windows 95
[ Self Reported ] Subsystem version 4.0 agrees with operating system version of Windows 95
Compiled: 2002-06-06 11:25:18 UTC
Compiler (one of):
         Visual C++ 4.2
         Visual C++ 5.0 (Visual Studio 97)
         Visual C++ 6.0 (Visual Studio 6.0)
Linker: Visual C++ 6.0 (Visual Studio 6.0)
Operating System: Windows 95
```

# Linker Versions

[source](https://web.archive.org/web/20231216085511/https://mariusbancila.ro/blog/2015/08/12/version-history-of-vc-mfc-and-atl/)

[wiki source under `Internal version numbering`](https://web.archive.org/web/20231216085732/https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B)

# Dll Versions

[source](https://web.archive.org/web/20231216075910/http://zuga.net/articles/vs-crt-runtime-library-versions/)

## License

Licensed under either of

 * Apache License, Version 2.0
   ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license
   ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

## Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be
dual licensed as above, without any additional terms or conditions.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!-- next-header -->
## [Unreleased] - ReleaseDate

## [0.1.0] - 2023-12-18

### Added
- Initial release

<!-- next-url -->
[Unreleased]: https://github.com/gtker/pe_info/compare/v0.1.0...HEAD
[0.1.0]: https://github.com/gtker/pe_info/tree/307e06dd5c8e3222f5c13fbd300a9fe04701dc39


src/compiler_version.rs
use chrono::{DateTime, NaiveDate, NaiveDateTime, NaiveTime, Utc};
use std::fmt::Formatter;

#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
pub enum CompilerVersion {
    VisualCPP1_0,
    VisualCPP1_5,
    VisualCPP2_0,
    VisualCPP4_0,
    VisualCPP4_1,
    VisualCPP4_2,
    VisualCPP5_0,
    VisualCPP6_0,
    VisualStudioDotNet2002,
    VisualStudioDotNet2003,
    VisualStudio2005,
    VisualStudio2008,
    VisualStudio2010,
    VisualStudio2012,
    VisualStudio2013,
    VisualStudio2015,
    VisualStudio2017,
    VisualStudio2019,
    VisualStudio2022,
}

impl std::fmt::Display for CompilerVersion {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            CompilerVersion::VisualCPP1_0 => f.write_str("Visual C++ 1.0"),
            CompilerVersion::VisualCPP1_5 => f.write_str("Visual C++ 1.5"),
            CompilerVersion::VisualCPP2_0 => f.write_str("Visual C++ 2.0"),
            CompilerVersion::VisualCPP4_0 => f.write_str("Visual C++ 4.0"),
            CompilerVersion::VisualCPP4_1 => f.write_str("Visual C++ 4.1"),
            CompilerVersion::VisualCPP4_2 => f.write_str("Visual C++ 4.2"),
            CompilerVersion::VisualCPP5_0 => f.write_str("Visual C++ 5.0 (Visual Studio 97)"),
            CompilerVersion::VisualCPP6_0 => f.write_str("Visual C++ 6.0 (Visual Studio 6.0)"),
            CompilerVersion::VisualStudioDotNet2002 => f.write_str("Visual Studio .NET 2002"),
            CompilerVersion::VisualStudioDotNet2003 => f.write_str("Visual Studio .NET 2003"),
            CompilerVersion::VisualStudio2005 => f.write_str("Visual Studio 2005"),
            CompilerVersion::VisualStudio2008 => f.write_str("Visual Studio 2008"),
            CompilerVersion::VisualStudio2010 => f.write_str("Visual Studio 2010"),
            CompilerVersion::VisualStudio2012 => f.write_str("Visual Studio 2012"),
            CompilerVersion::VisualStudio2013 => f.write_str("Visual Studio 2013"),
            CompilerVersion::VisualStudio2015 => f.write_str("Visual Studio 2015"),
            CompilerVersion::VisualStudio2017 => f.write_str("Visual Studio 2017"),
            CompilerVersion::VisualStudio2019 => f.write_str("Visual Studio 2019"),
            CompilerVersion::VisualStudio2022 => f.write_str("Visual Studio 2022"),
        }
    }
}

fn date_to_datetime(year: i32, month: u32, day: u32) -> DateTime<Utc> {
    DateTime::from_naive_utc_and_offset(
        NaiveDateTime::new(
            match NaiveDate::from_ymd_opt(year, month, day) {
                None => panic!("invalid date"),
                Some(e) => e,
            },
            NaiveTime::MIN,
        ),
        Utc,
    )
}

impl CompilerVersion {
    pub fn release_date(&self) -> DateTime<Utc> {
        match self {
            CompilerVersion::VisualCPP1_0 => date_to_datetime(1993, 2, 22),
            CompilerVersion::VisualCPP1_5 => date_to_datetime(1993, 12, 1),
            CompilerVersion::VisualCPP2_0 => date_to_datetime(1994, 9, 16),
            CompilerVersion::VisualCPP4_0 => date_to_datetime(1995, 12, 11),
            CompilerVersion::VisualCPP4_1 => date_to_datetime(1996, 2, 27),
            CompilerVersion::VisualCPP4_2 => date_to_datetime(1996, 7, 22),
            CompilerVersion::VisualCPP5_0 => date_to_datetime(1997, 3, 19),
            CompilerVersion::VisualCPP6_0 => date_to_datetime(1998, 9, 2),
            CompilerVersion::VisualStudioDotNet2002 => date_to_datetime(2002, 2, 13),
            CompilerVersion::VisualStudioDotNet2003 => date_to_datetime(2003, 4, 24),
            CompilerVersion::VisualStudio2005 => date_to_datetime(2005, 11, 7),
            CompilerVersion::VisualStudio2008 => date_to_datetime(2007, 11, 19),
            CompilerVersion::VisualStudio2010 => date_to_datetime(2010, 4, 12),
            CompilerVersion::VisualStudio2012 => date_to_datetime(2012, 9, 12),
            CompilerVersion::VisualStudio2013 => date_to_datetime(2013, 10, 17),
            CompilerVersion::VisualStudio2015 => date_to_datetime(2015, 6, 20),
            CompilerVersion::VisualStudio2017 => date_to_datetime(2017, 3, 7),
            CompilerVersion::VisualStudio2019 => date_to_datetime(2019, 4, 2),
            CompilerVersion::VisualStudio2022 => date_to_datetime(2021, 11, 8),
        }
    }

    pub fn newest_possible(datetime: DateTime<Utc>) -> Self {
        let mut newest = Self::VisualCPP1_0;

        for v in Self::values() {
            if v.release_date() > datetime {
                return newest;
            } else {
                newest = *v;
            }
        }

        newest
    }

    fn values() -> &'static [Self] {
        &[
            Self::VisualCPP1_0,
            Self::VisualCPP1_5,
            Self::VisualCPP2_0,
            Self::VisualCPP4_0,
            Self::VisualCPP4_1,
            Self::VisualCPP4_2,
            Self::VisualCPP5_0,
            Self::VisualCPP6_0,
            Self::VisualStudioDotNet2002,
            Self::VisualStudioDotNet2003,
            Self::VisualStudio2005,
            Self::VisualStudio2008,
            Self::VisualStudio2010,
            Self::VisualStudio2012,
            Self::VisualStudio2013,
            Self::VisualStudio2015,
            Self::VisualStudio2017,
            Self::VisualStudio2019,
            Self::VisualStudio2022,
        ]
    }
}
